### ISSUE TITLE
Macro derive fails when a non-std write macro is in scope

### ISSUE THREAD

**Issue by konstin** [dtolnay/thiserror#239]
When we have our own [custom `write!` macro](https://github.com/astral-sh/ruff/blob/528bf2df3adb731c1d1e8021ed931e8a2c152037/crates/ruff_formatter/src/macros.rs#L69-L75) with a different declaration than the std one in scope, the `Error` derive fails ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0bda49d06950ba4b815d1b3f4ad75c71)):

```rust
use thiserror::Error;
use std::io;

/// Our custom write macro
#[macro_export]
macro_rules! write {
    ($dst:expr, [$($arg:expr),+ $(,)?]) => {{
        let result = $dst.write_fmt($crate::format_args!($($arg),+));
        result
    }}
}

#[derive(Error, Debug)]
pub enum FormatModuleError {
    #[error("io error: {0}")]
    IoError(io::Error),
}
```

```rust
error: no rules expected the token `"io error: {field__0}"`
  --> src/lib.rs:16:13
   |
6  | macro_rules! write {
   | ------------------ when calling this macro
...
16 |     #[error("io error: {0}")]
   |             ^^^^^^^^^^^^^^^ no rules expected this token in macro call
   |
note: while trying to match `[`
  --> src/lib.rs:7:17
   |
7  |     ($dst:expr, [$($arg:expr),+ $(,)?]) => {{
   |                 ^
```

**Comment by dtolnay** [comment]
Good catch. Fixed in 1.0.42.

**Comment by konstin** [comment]
Thanks!
